File: Fibonacci number.cpp
//Problem Link:  https://leetcode.com/problems/fibonacci-number/



class Solution {
public:
    int dp[31];
    int fib(int N) {
        if(N<0)
            return 0;
        if(N==1)
            return 1;
        if(dp[N]!=0)
            return dp[N];
        return dp[N] = fib(N-1)+fib(N-2);
    }
};

File: Fibonacci number.js
//https://leetcode.com/problems/fibonacci-number/


/**
 * @param {number} N
 * @return {number}
 */
let dp = new Array(31);

function fibo(N) {
    if(dp[N]!=undefined)
        return dp[N];
    return dp[N]=fibo(N-1)+fibo(N-2);
}

var fib = function(N) {
    dp[0]=0,dp[1]=1;
    return fibo(N);
};